File: TeamCode/src/main/java/teamcode/teleop/Teleop.java
package teamcode.teleop;

import static java.lang.Double.max;
import static java.lang.Double.min;

import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import teamcode.robot.core.Alliance;
import teamcode.robot.core.RobotHardware;
import teamcode.robot.subsystems.Shooter;
import teamcode.robot.threads.MovementThread;
import teamcode.robot.threads.TurretThread;
import teamcode.robot.threads.VisionThread;
import teamcode.threading.ThreadedOpMode;

/**
 * Example usage of ThreadedOpMode.
 * This demonstrates how to use threading in your robot OpMode.
 * Remove @Disabled annotation to enable this OpMode.
 */
@TeleOp(name = "Teleop", group = "Teleop")
public class Teleop extends ThreadedOpMode {
    private MovementThread movementThread;
    private TurretThread turretThread;
    private VisionThread visionThread;
    
    @Override
    protected void initializeThreads() {
        // Create and configure threads
        movementThread = new MovementThread();
        turretThread = new TurretThread();
        visionThread = new VisionThread();
        
        // Add threads to manager
        threadManager.addThread(movementThread);
        threadManager.addThread(visionThread);
        threadManager.addThread(turretThread);
    }

    @Override
    protected void runInit() {
        Alliance alliance= Alliance.RED;

        while(opModeInInit()){
            telemetry.addData("Status", "In Init");


            if (gamepad1.xWasPressed()){
                alliance = Alliance.BLUE;
            }
            if (gamepad1.bWasPressed()){
                alliance = Alliance.RED;
            }
            telemetry.update();
        }

        RobotHardware.setAlliance(alliance);
    }
    
    // State variables for button press detection
    private boolean firstPressA = false;
    private boolean firstPressB = false;
    private boolean firstPressRightBumper = false;
    private boolean firstPressLeftBumper = false;
    
    @Override
    protected void onStart() {
        // Enable turret control when OpMode starts
        turretThread.enable();
    }
    
    @Override
    protected void mainLoop() {
        // Main loop iteration - called repeatedly by ThreadedOpMode
        // Telemetry staging and updates are handled automatically!
        
        // ===== MOVEMENT CONTROL =====
        // Update movement thread with gamepad1 input
        movementThread.setDriveInput(gamepad1);
        
        // Toggle precision mode with left trigger (reduces speed to 30%)
        if (gamepad1.left_trigger > 0.5) {
            movementThread.setSpeedMultiplier(0.3);
        } else {
            movementThread.setSpeedMultiplier(1.0);
        }
        
        // ===== TURRET CONTROL =====
        // Turret automatically tracks targets using vision-based PID
        firstPressA = gamepad1.a && !firstPressA;
        firstPressB = gamepad1.b && !firstPressB;
        firstPressRightBumper = gamepad1.right_bumper && !firstPressRightBumper;
        firstPressLeftBumper = gamepad1.left_bumper && !firstPressLeftBumper;

        if (firstPressA) {
            turretThread.enable();
        }

        if (firstPressB) {
            turretThread.disable();
        }

        RobotHardware.turretShooterRightMotor.setCurrentPosition(gamepad1.right_trigger);
        RobotHardware.turretShooterLeftMotor.setCurrentPosition(gamepad1.right_trigger);

        telemetry.addData("Shooter Power", gamepad1.right_trigger);

        // ===== TELEMETRY =====
        // Just use telemetry.addData() - namespace and updates are handled automatically!
        telemetry.addData("Runtime", String.format("%.2f s", runtime.seconds()));
        telemetry.addData("Loop Time", String.format("%.0f ms", runtime.milliseconds()));
        
        // Add servo position if available
        if (RobotHardware.kickerServo != null) {
            telemetry.addData("Kicker Servo Pos", String.format("%.2f", RobotHardware.kickerServo.getPosition()));
        }
    }
    
    @Override
    protected void cleanup() {
        // Disable threads before stopping
        movementThread.disable();
        turretThread.disable();
    }
}

File: TeamCode/src/main/java/teamcode/teleop/Teleop.kt
package teamcode.teleop

import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import teamcode.robot.core.Alliance
import teamcode.robot.core.RobotHardware
import teamcode.robot.core.state.RobotState
import teamcode.robot.core.state.RobotStateMachine
import teamcode.robot.subsystems.ColorSensorSubsystem
import teamcode.robot.subsystems.IntakeSubsystem
import teamcode.robot.subsystems.KickerSubsystem
import teamcode.robot.subsystems.MovementSubsystem
import teamcode.robot.subsystems.ShooterState
import teamcode.robot.subsystems.ShooterSubsystem
import teamcode.robot.subsystems.SpindexerSubsystem
import teamcode.robot.subsystems.TurretState
import teamcode.robot.subsystems.TurretSubsystem
import teamcode.robot.subsystems.VisionSubsystem
import teamcode.threading.ThreadedOpMode
import teamcode.commands.ShootAndTurn
import teamcode.commands.TriggerKicker
import teamcode.commands.SpindexerSpin
import teamcode.robot.command.execute
import teamcode.robot.subsystems.KickerState
import teamcode.robot.subsystems.VisionPipeline
import kotlin.math.min

/**
 * Main teleop OpMode.
 * Clean architecture with automatic command scheduling.
 */
@TeleOp(name = "Teleop", group = "Teleop")
class Teleop : ThreadedOpMode() {
    
    // Subsystems
    private lateinit var movementSubsystem: MovementSubsystem
    private lateinit var turretSubsystem: TurretSubsystem
    private lateinit var visionSubsystem: VisionSubsystem
    private lateinit var shooterSubsystem: ShooterSubsystem
    private lateinit var kickerSubsystem: KickerSubsystem
    private lateinit var spindexerSubsystem: SpindexerSubsystem
    private lateinit var intakeSubsystem: IntakeSubsystem
    private lateinit var colorSensorSubsystem: ColorSensorSubsystem

    
    override fun initOpMode() {
        // Initialize subsystems
        movementSubsystem = MovementSubsystem()
        turretSubsystem = TurretSubsystem()
        visionSubsystem = VisionSubsystem()
        shooterSubsystem = ShooterSubsystem()
        kickerSubsystem = KickerSubsystem()
        spindexerSubsystem = SpindexerSubsystem()
        intakeSubsystem = IntakeSubsystem()
        colorSensorSubsystem = ColorSensorSubsystem()

            telemetry.addData("Status", "In Init")

    }

    override fun onStart() {
        turretSubsystem.enable()
//        turretSubsystem.enterManualMode()
        visionSubsystem.setPipeline(VisionPipeline.RED)
    }
    
    override fun mainLoop() {
        when (RobotStateMachine.getState()){
            RobotState.IDLE -> {
                if (gamepad2Ex.rightTrigger.value>0.05){
                    RobotStateMachine.transitionTo(RobotState.SHOOTING)
                }
            }
            RobotState.SHOOTING -> {
                if (gamepad2Ex.rightTrigger.value<0.05 && kickerSubsystem.currentState== KickerState.IDLE){
                    RobotStateMachine.transitionTo(RobotState.IDLE)
                    return
                }
                // Schedule shoot command when B button is pressed
                if (gamepad2Ex.b.wasPressed()){
                    TriggerKicker().execute()
                }
                if (gamepad2Ex.dpadUp.wasPressed()){
                    ShootAndTurn().execute()
                }
            }
            RobotState.INTAKING -> {

            }


        }

        // Spindexer control - schedule cycle command when A button is pressed
        if (gamepad2Ex.a.wasPressed()){
            SpindexerSpin(1).execute()
        }

//        // Turret control - execute manual turret commands based on bumper input
//        if (turretSubsystem.currentState == TurretState.MANUAL){
//            if (gamepad2Ex.rightBumper.value){
//                turretSubsystem
//            } else if(gamepad2Ex.leftBumper.value){
//                ManualTurretCommand(turretSubsystem, -1.0).execute()
//            } else {
//                ManualTurretCommand(turretSubsystem, 0.0).execute()
//            }
//        }

        // Drive control
        movementSubsystem.setDriveInput(gamepad1)


        if(gamepad2Ex.dpadLeft.wasPressed()) {
            RobotHardware.turretTurnServo.set(RobotHardware.turretTurnServo.get()+0.1)
        }
        if(gamepad2Ex.dpadRight.wasPressed()) {
            RobotHardware.turretTurnServo.set(RobotHardware.turretTurnServo.get()-0.1)
        }
        // ===== TELEMETRY =====
        robotTelemetry.addData("Servo",RobotHardware.turretTurnServo.get())
        robotTelemetry.addData("Robot State", getState().name)
        robotTelemetry.addData("Runtime", runtime.seconds())
    }
    
    override fun cleanup() {
        movementSubsystem.disable()
        turretSubsystem.disable()
    }
}
